cupy/math/sumprod.py

```diff
@@ -172,31 +169,38 @@ def _cum_core(a, axis, dtype, out, kern, batch_kern):
     else:
         return _proc_as_batch(batch_kern, out, axis=axis)
 
-    pos = 1
-    while pos < out.size:
-        kern(pos, out, size=out.size)
-        pos <<= 1
+    kern(out.size, out, size=out.size)
     return out
 
 
 _cumsum_batch_kern = core.ElementwiseKernel(
-    'int64 pos, int64 batch', 'raw T x',
+    'int64 size, int64 batch', 'raw T x',
     '''
-    ptrdiff_t b = i % batch;
-    ptrdiff_t j = i / batch;
-    if (j & pos) {
-      const ptrdiff_t dst_index[] = {j, b};
-      const ptrdiff_t src_index[] = {j ^ pos | (pos - 1), b};
-      x[dst_index] += x[src_index];
+    long long pos = 1;
+    while(pos < size) {
+      ptrdiff_t b = i % batch;
+      ptrdiff_t j = i / batch;
+      if (j & pos) {
+        const ptrdiff_t dst_index[] = {j, b};
+        const ptrdiff_t src_index[] = {j ^ pos | (pos - 1), b};
+        x[dst_index] += x[src_index];
+      }
+      pos <<= 1;
+      __syncthreads();
```

Comment:
`__synchthreads` does not synchronize two threads belong to a different block. So it seems to occur data races for large input.

Reply:
Good point, apparently to synchronize the grid compute capability >= 6 is needed so we need to rewrite the kernel to avoid such data races...